apps/yggdrasil/priv/repo/migrations/20160229163000_create_role_resource.exs

```diff
@@ -0,0 +1,17 @@
+defmodule Yggdrasil.Repo.Migrations.CreateRoleResource do
+  use Ecto.Migration
+
+  def change do
+    create table(:roles_resources) do
+      add :role_id, references(:roles, on_delete: :nothing)
+      add :resource_id, references(:resources, on_delete: :nothing)
+      add :permission_id, references(:permissions, on_delete: :nothing)
+
+      timestamps
+    end
+    create index(:roles_resources, [:role_id])
+    create index(:roles_resources, [:resource_id])
+    create index(:roles_resources, [:permission_id])
```

Comment:
No I'm talking about a primary key for the table. Role+Resource+Permission together could be the primary key as those three identify a unique record. It just so happens there is no additional information tied to it.

The primary key usually is what dictates the order (I believe in postgres even that is customizable) that the records are stored in. If Role+Resource+Permission was the key then the records would be stored in that order, naturally grouping all of a roles permissions together. You wouldn't need an index to speed up lookup by role, the database should be able to optimize the lookup based on the primary key information.

The downside of this is when you insert a new record into the database, it might have to shuffle records around to make space on a page. When you use an auto-incrementing integer for the id, it means the new record is always added to the end of the table. You wouldn't have the issue of the database having to move data around to insert a record in the middle of a page, but there is no natural grouping of the records... well except by insert order.